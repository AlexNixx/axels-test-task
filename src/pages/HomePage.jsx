import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { PropertyItem } from '../components/PropertyItem';
import { fetchProperties } from '../redux/ducks/properties';
import {
    StyledCol,
    StyledContainer,
    StyledRow
} from '../styles/pages/HomePageStyles';

export const HomePage = () => {
    const dispatch = useDispatch();
    const properties = useSelector(state => state.propertiesReducer.properties);
    const error = useSelector(state => state.propertiesReducer.error);
    const loading = useSelector(state => state.propertiesReducer.loading);

    useEffect(() => {
        dispatch(fetchProperties());
    }, [dispatch]);

    if (loading) return <p>Loading...</p>;

    if (error) return <p>{error}</p>;

    return (
        <StyledContainer>
            <StyledRow>
                {properties?.map(property => (
                    <StyledCol xs={12} md={6} lg={4} xxl={3} key={property.id}>
                        <PropertyItem {...property} />
                    </StyledCol>
                ))}
            </StyledRow>
        </StyledContainer>
    );
};

HomePage.propTypes = {};
