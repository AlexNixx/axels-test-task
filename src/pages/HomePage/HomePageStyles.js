import styled from 'styled-components';
import { Col, Container, Row } from 'react-bootstrap';

export const StyledContainer = styled(Container)`
    padding: 1rem;
`;

export const StyledRow = styled(Row)`
    row-gap: 2rem;
`;

export const StyledCol = styled(Col)`
    display: flex;
    justify-content: center;
`;
