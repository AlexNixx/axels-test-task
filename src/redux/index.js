import { applyMiddleware, combineReducers, createStore } from 'redux';
import createSagaMiddleware from 'redux-saga';

import propertiesReducer from './ducks/properties';
import propertyReducer from './ducks/property';

import { rootWatcher } from './rootWatcher';

const sagaMiddleware = createSagaMiddleware();

const rootReducer = combineReducers({
    propertiesReducer,
    propertyReducer
});

export const store = createStore(rootReducer, applyMiddleware(sagaMiddleware));

sagaMiddleware.run(rootWatcher);
