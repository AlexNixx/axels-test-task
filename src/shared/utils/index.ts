export * from './test/renderWithProviders';
export * from './test/showLocation';
export * from './convertPrice/convertPrice';
