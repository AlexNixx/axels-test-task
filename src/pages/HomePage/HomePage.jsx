import {PropertyItem} from "../../components/PropertyItem/PropertyItem";
import {mockData} from "../../mockData/mockData";
import {StyledCol, StyledContainer, StyledRow} from "./HomePageStyles";

export const HomePage = () => {
    return (
        <StyledContainer>
            <StyledRow xs={1} md={2} lg={3} xl={4}>
                {mockData.map((property) => (
                    <StyledCol key={property.id}>
                        <PropertyItem {...property} />
                    </StyledCol>
                ))}
            </StyledRow>
        </StyledContainer>
    );
};
