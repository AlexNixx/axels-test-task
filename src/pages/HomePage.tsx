import { useEffect } from 'react';

import { PropertyItem } from '../components/PropertyItem';

import { fetchProperties } from '../redux/ducks/properties';
import { useAppDispatch, useAppSelector } from '../redux';
import {
    StyledCol,
    StyledContainer,
    StyledRow
} from '../styles/pages/HomePageStyles';

export const HomePage = () => {
    const dispatch = useAppDispatch();
    const properties = useAppSelector(
        state => state.propertiesReducer.properties
    );
    const error = useAppSelector(state => state.propertiesReducer.error);
    const loading = useAppSelector(state => state.propertiesReducer.loading);

    useEffect(() => {
        dispatch(fetchProperties());
    }, [dispatch]);

    if (loading) return <p>Loading...</p>;

    if (error) return <p>{error}</p>;

    return (
        <StyledContainer>
            <StyledRow>
                {properties?.map(property => (
                    <StyledCol xs={12} md={6} lg={4} xxl={3} key={property.id}>
                        <PropertyItem {...property} />
                    </StyledCol>
                ))}
            </StyledRow>
        </StyledContainer>
    );
};
