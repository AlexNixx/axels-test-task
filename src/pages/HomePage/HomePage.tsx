import { useEffect } from 'react';

import { PropertyItem } from '../../components/PropertyItem';

import { fetchProperties } from '../../redux/ducks/properties';
import { useAppDispatch, useAppSelector } from '../../redux';
import { StyledCol, HomeContainer, StyledRow } from '../../styles/pages';

export const HomePage = () => {
    const dispatch = useAppDispatch();
    const properties = useAppSelector(
        state => state.propertiesReducer.properties
    );
    const error = useAppSelector(state => state.propertiesReducer.error);
    const loading = useAppSelector(state => state.propertiesReducer.loading);

    useEffect(() => {
        dispatch(fetchProperties());
    }, [dispatch]);

    if (loading) return <p data-testid='loadingText'>Loading...</p>;

    if (error) return <p data-testid='errorText'>{error}</p>;

    return (
        <HomeContainer data-testid='homePage'>
            <StyledRow>
                {properties?.map(property => (
                    <StyledCol xs={12} md={6} lg={4} xxl={3} key={property.id}>
                        <PropertyItem {...property} />
                    </StyledCol>
                ))}
            </StyledRow>
        </HomeContainer>
    );
};
