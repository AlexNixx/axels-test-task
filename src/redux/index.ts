import { applyMiddleware, combineReducers, createStore } from 'redux';
import { TypedUseSelectorHook, useSelector, useDispatch } from 'react-redux';
import createSagaMiddleware from 'redux-saga';

import propertiesReducer from './ducks/properties';
import propertyReducer from './ducks/property';
import { rootWatcher } from './rootWatcher';

const sagaMiddleware = createSagaMiddleware();

const rootReducer = combineReducers({
    propertiesReducer,
    propertyReducer
});

export const store = createStore(rootReducer, applyMiddleware(sagaMiddleware));

sagaMiddleware.run(rootWatcher);

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
