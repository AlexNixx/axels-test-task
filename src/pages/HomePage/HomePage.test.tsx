import { screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { HomePage } from './HomePage';

import { renderWithProviders, mockProperties } from '../../utils';
import {
    fetchProperties,
    setError,
    setProperties
} from '../../redux/ducks/properties';
import { store } from '../../redux';

const mockDispatch = jest.fn();

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useDispatch: () => mockDispatch
}));

describe('Test Home Page', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    test('should be display loading state', () => {
        store.dispatch(fetchProperties());

        renderWithProviders(<HomePage />, {});

        expect(mockDispatch).toHaveBeenCalledTimes(1);
        expect(screen.getByTestId('loadingText')).toBeInTheDocument();
    });

    test('should be display error state', async () => {
        const errorMessage = 'An error occurred';

        store.dispatch(setError(errorMessage));
        renderWithProviders(<HomePage />, {});

        expect(mockDispatch).toHaveBeenCalledTimes(1);
        expect(screen.getByTestId('errorText')).toBeInTheDocument();
        expect(screen.getByTestId('errorText')).toContainHTML(errorMessage);
    });

    test('should be display properties cards', async () => {
        store.dispatch(setProperties(mockProperties));
        const { container } = renderWithProviders(<HomePage />, {});

        expect(screen.getByText(/Test Property/)).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });
});
