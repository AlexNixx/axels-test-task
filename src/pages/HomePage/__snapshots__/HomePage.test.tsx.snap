// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Home Page should be display properties cards 1`] = `
<div>
  <div
    class="sc-jSwlEQ ePyzur container"
    data-testid="homePage"
  >
    <div
      class="sc-gKHVLF eEIIgr row"
    >
      <div
        class="sc-iBAaJG clwkyP col-xxl-3 col-lg-4 col-md-6 col-12"
      >
        <a
          data-testid="propertyLink"
          href="/property/1"
        >
          <div
            class="sc-bcPKhP bZwkZu card"
            data-testid="propertyCard"
          >
            <img
              alt="property Test Property"
              class="card-img-top sc-grXZZQ dsaGvy"
              src="test.jpg"
            />
            <div
              class="card-body"
            >
              <div
                class="card-title h5"
                data-testid="cardTitle"
              >
                Test Property
              </div>
              <div
                class="card-subtitle h6"
                data-testid="cardPrice"
              >
                $100.00
              </div>
              <p
                class="card-text"
                data-testid="cardAddress"
              >
                123 Test St
              </p>
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;
